<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Transcriptomic TimeSeries (longitudinal) analysis • TimeSeriesAnalysis</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Transcriptomic TimeSeries (longitudinal) analysis">
<meta property="og:description" content="A pipeline to analyse and interpret longitudinal transcriptomic data from both RNAseq and microarray origins. The pipeline performs differential gene expression, clustering, and plots usefull figures to aid in interpretation of identified clusters of significant genes.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">TimeSeriesAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">01.01.01</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Examples
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/TS_analysis_PBMC.html">PBMC analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Ylefol/TimeSeriesAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="timeseriesanalysis">TimeSeriesAnalysis<a class="anchor" aria-label="anchor" href="#timeseriesanalysis"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p>TiSA: TimeSeriesAnalysis - A transcriptomic analysis tool for both RNA sequencing and microarray data</p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>TimeSeriesAnalysis (TiSA) is an analysis and visualization package for RNAseq and microarray data. TS extracts significant genes from time course transcriptomic data by performing differential gene expression on both the conditional and temporal axes. It then employs partitioning algorithm based on recursive thresholding (PART) clustering to identify small genomic clusters of relevance, followed by running the clusters through gprofiler to reveal the biological relevance of each cluster.</p>
</div>
<div class="section level2">
<h2 id="ts-performs">TS performs:<a class="anchor" aria-label="anchor" href="#ts-performs"></a>
</h2>
<ul>
<li>Data normalization and processing</li>
<li>PCA plots</li>
<li>Differential gene expression (conditional and temporal)</li>
<li>PART clustering</li>
<li>Heatmaps for both differential expression summary and clustering results</li>
<li>Trajectory of identified clusters</li>
<li>Gprofiler (functional enrichment) analysis of clusters</li>
<li>Dotplots and MDS plots of Gprofiler results</li>
<li>Nearest ancestor clustering of GOs</li>
<li>GO ancestor queries</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the stable version of TimeSeriesAnalysis like so:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remotes/man/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"Ylefol/TimeSeriesAnalysis@master"</span><span class="op">)</span></span></code></pre></div>
<p>The development version can be installed by replacing ‘master’ with ‘dev’.</p>
<p>Devtools is required for the installation of this package. To install Devtools, visit the <a href="https://www.r-project.org/nosvn/pandoc/devtools.html" class="external-link">Devtools main page</a></p>
<p>Certain bioconductor packages will have to be installed before installation of TimeSeriesAnalysis. The code snippet below gives the method along with some of the packages that should be installed via this method.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>bio_pkgs <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'DESeq2'</span>,<span class="st">'GOSemSim'</span>,<span class="st">'GO.db'</span>,<span class="st">'limma'</span>,<span class="st">'ComplexHeatmap'</span>,</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="st">'AnnotationDbi'</span>,<span class="st">'GenomeInfoDb'</span>,<span class="st">'GenomicRanges'</span>,<span class="st">'Biobase'</span>,<span class="st">'S4Vectors'</span>,</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="st">'BioCGenerics'</span>,<span class="st">'MatrixGenerics'</span>,<span class="st">'IRanges'</span>,<span class="st">'BiocFileCache'</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="st">'SummarizedExperiment'</span>,<span class="st">'org.Hs.eg.db'</span>,<span class="st">'org.Mm.eg.db'</span>,<span class="st">'org.Ce.eg.db'</span>)</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(bio_pkgs)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="rmarkdown-format">Rmarkdown format<a class="anchor" aria-label="anchor" href="#rmarkdown-format"></a>
</h2>
<p>TimeSeriesAnalysis was developed with user-friendliness in mind, the core script of the package is a Rmarkdown file with explanations of what each code block performs. Computationally intensive code blocks save their work in order to avoid loss of computation time. TimeSeriesAnalysis comes with a test dataset. Users are recommended to first run the package with the test data to then modify the Rmarkdown script with the necessary information for their purposes (input files, organism, genes of interest etc…). The rmarkdown script and the expected result from an example run can be found and downloaded from this repository, they are located in the ‘rmarkdown_method’ folder. If users prefer a script approach, the equivalent of the rmarkdown report is provided in two scripts, one for the computation tasks and a second for the analyses. These two scripts are found in the ‘script_method’ folder.</p>
</div>
<div class="section level2">
<h2 id="example-data">Example Data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h2>
<p>Three different datasets have been used to test this pipeline, one of which has been preserved as an example datasets for the vignettes of this pipeline. More information on the other datasets can be viewed through the publications listed below.</p>
<p>The PBMC dataset is a time series experiment with three time points that explores and compares three AICDA/AID (Activation Induced Cytidine DeAminase) activation cocktails. The experiment seeked to identify which activation cocktail properly activated AID through both the expression of the AID gene and the activation of class switch recombination.</p>
<p>The MURINE dataset is another AID related time series experiment, where mice were generated and one was given an AID activation cocktail. The experiment contains one replicate per condition and 10 timepoints. Due to only having one replicate per condition per timepoint, subsequent timepoints were combined to create an intermediary timepoint with two replicates in each condition.</p>
<p>The Celegans dataset is an ageing related experiment. Two celegans models were used: BY372 and N2. The first being a parkinsons model and the second being a wild type. Each model was split into two groups, one that would be fed krill oil and the other not (control). Each experiment has three timepoints: Days 1, 3, and 6.</p>
</div>
<div class="section level2">
<h2 id="tutorial">Tutorial<a class="anchor" aria-label="anchor" href="#tutorial"></a>
</h2>
<p>A tutorial using the PBMC data can be found within the <a href="https://ylefol.github.io/TimeSeriesAnalysis/" class="external-link">pipelines website</a>. Additionally the documentation can be found in the references tab.</p>
<div class="section level3">
<h3 id="set-up-of-a-sample-file">Set-up of a sample file<a class="anchor" aria-label="anchor" href="#set-up-of-a-sample-file"></a>
</h3>
<p>In order for the pipeline to run properly, a sample file must be provided. Certain requirements ust be met. Column names should be as indicated below, all spaces should be avoided, timepoints should be numerical. It is also important that the sample column contains the same names as the count files provided. Lastly, replicates should always be followed by a ’_#’ as this numbering will be utilized for the organization of the replicates within certain visualizations. Note that in the below example the three files have the same name (ex: C_1). That is to represent the ‘C_1’ sample at its three different time points.</p>
<table class="table">
<thead><tr class="header">
<th align="left">sample</th>
<th align="left">group</th>
<th align="left">replicate</th>
<th align="right">timepoint</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">s-37</td>
<td align="left">control</td>
<td align="left">C_1</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">s-38</td>
<td align="left">control</td>
<td align="left">C_1</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">s-39</td>
<td align="left">control</td>
<td align="left">C_1</td>
<td align="right">9</td>
</tr>
<tr class="even">
<td align="left">s-52</td>
<td align="left">IgM</td>
<td align="left">IgM_1</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">s-53</td>
<td align="left">IgM</td>
<td align="left">IgM_1</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">s-54</td>
<td align="left">IgM</td>
<td align="left">IgM_1</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="left">s-77</td>
<td align="left">control</td>
<td align="left">C_2</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">s-78</td>
<td align="left">control</td>
<td align="left">C_2</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">s-79</td>
<td align="left">control</td>
<td align="left">C_2</td>
<td align="right">9</td>
</tr>
<tr class="even">
<td align="left">s-92</td>
<td align="left">IgM</td>
<td align="left">IgM_2</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">s-93</td>
<td align="left">IgM</td>
<td align="left">IgM_2</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">s-94</td>
<td align="left">IgM</td>
<td align="left">IgM_2</td>
<td align="right">9</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="rnaseq-based-data">RNAseq based data<a class="anchor" aria-label="anchor" href="#rnaseq-based-data"></a>
</h3>
<p>TiSA utilizes DESeq2 to process RNAseq data. It therefore requires that the data be in a raw count format. The pipeline can take in multiple count files within a folder provided that the file names match the sample names given in the sample file. These individual files should be tab delineated, contain only two columns (gene ID and associated count), there should be no column header. Additionally each file should end with a ‘.counts’.</p>
<p>As an alternative, all counts may be within the same text file. In this case the first column should be labeled as ‘gene_id’, and each subsequent column should contain the same sample name as the names found in the sample file. This file should end with ‘.txt’.</p>
<p>In either case, the given information should be counts (round numbers).</p>
</div>
</div>
<div class="section level2">
<h2 id="microarray-based-data">Microarray based data<a class="anchor" aria-label="anchor" href="#microarray-based-data"></a>
</h2>
<p>For microarray data, a streamlined method is in the works. Currently microarray data needs to be inputed as a Elist, specifically an E list saved as a rds object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_path_data</span><span class="op">&lt;-</span><span class="st">'data/micro_arr/my_limma_dta.rds'</span></span>
<span><span class="va">my_path_sample_dta</span><span class="op">&lt;-</span><span class="st">'data/micro_arr/sample_file.csv'</span></span>
<span></span>
<span><span class="co">#Set-up time series object parameters</span></span>
<span><span class="va">diff_exp_type</span><span class="op">&lt;-</span><span class="st">'limma'</span> </span></code></pre></div>
<p>The rds file given to ‘my_path_data’ contains the Elist produced by limma processing of microarray data. It is also important to set the differential gene expression type (diff_exp_type) to ‘limma’.</p>
</div>
<div class="section level2">
<h2 id="using-the-pipeline">Using the pipeline<a class="anchor" aria-label="anchor" href="#using-the-pipeline"></a>
</h2>
<p>NOTE: You must still install the TimeSeriesAnalysis package as defined above</p>
<p>Since the pipeline’s source code is available, individuals are free to take the code and adapt it as needed. However for users who would like the easiest approach to use the pipeline as is, it is recommended to clone or download this github repository to your local computer and work from the Rmarkdown method folder within the repository. Cloning or downloading can be done from the <a href="https://github.com/Ylefol/TimeSeriesAnalysis" class="external-link">TimeSeriesAnalysis github page</a> by using the green button called ‘code’. From there, the Rmarkdown file can be edited.</p>
<p>To edit the file, it is recommended to use rstudio, which can be downloaded <a href="https://www.rstudio.com/products/rstudio/download/" class="external-link">here</a>. Note that it is recommended to download the free version of RStudio Desktop, not RStudio Server.</p>
<p>The only code block of code is the parameter set-up block (2nd) as well as the title at the very top of the Rmarkdown document.</p>
</div>
<div class="section level2">
<h2 id="launching-the-pipeline">Launching the pipeline<a class="anchor" aria-label="anchor" href="#launching-the-pipeline"></a>
</h2>
<p>The pipeline can be launched from rstudio itself by using the ‘knit’ button, or it can be launched from a command line using the following command from the TimeSeriesAnalysis repository:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">"rmarkdown::render('rmarkdown_method/TS_analysis.Rmd',output_file='TS_analysis.html')"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="publication">Publication<a class="anchor" aria-label="anchor" href="#publication"></a>
</h2>
<p>A manuscript detailing the TimeSeriesAnalysis pipeline published with NAR genomics and bioinformatics. <a href="https://academic.oup.com/nargab/article/5/1/lqad020/7069286" class="external-link">TiSA: TimeSeriesAnalysis—a pipeline for the analysis of longitudinal transcriptomics data</a> The version/DOI of this pipeline used within the manuscript is 10.5281/zenodo.7616032.</p>
<p>A ageing related study was performed in part through the use of TiSA. <a href="https://www.aging-us.com/article/204375/text" class="external-link">Krill oil protects dopaminergic neurons from age-related degeneration through temporal transcriptome rewiring and suppression of several hallmarks of aging</a></p>
<p>A longitudinal COVID-19 study was performed using TiSA. <a href="https://www.nature.com/articles/s41598-023-37606-y" class="external-link">Longitudinal whole blood transcriptomic analysis characterizes neutrophil activation and interferon signaling in moderate and severe COVID-19</a></p>
</div>
<div class="section level2">
<h2 id="additional-information">Additional information<a class="anchor" aria-label="anchor" href="#additional-information"></a>
</h2>
<p>The ‘clusterGenomics’ package was no longer maintained on CRAN and therefore the necessary scripts were brought over to this package for it’s implementation without the need to download the clusterGenomics package from source. Within this package, in the ‘R/clusterGenomics.R’ code was written by the authors of clusterGenomics. It was imported to TimeSeriesAnalysis by the authors of TimeSeriesAnalysis.</p>
</div>
<div class="section level2">
<h2 id="contact-information">Contact information<a class="anchor" aria-label="anchor" href="#contact-information"></a>
</h2>
<p>If you have encountered an error or require assistance running the pipeline, feel free to open a github issue, or email me at <a href="mailto:yohanl@medisin.uio.no" class="email">yohanl@medisin.uio.no</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Ylefol/TimeSeriesAnalysis/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing TimeSeriesAnalysis</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yohan Lefol <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Yohan Lefol.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
