<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create gprofiler dotplots — GO_dotplot_wrapper • TimeSeriesAnalysis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Create gprofiler dotplots — GO_dotplot_wrapper"><meta property="og:description" content="Wrapper function for the creation of a dotplot which summarises the gprofiler findings
for a specified ontology.
The function reads the results for a location. It will automatically search for the
'gprofiler_results/data_files' folder in the procided location. This is should be located
in the TS_results folder, but can change based on the user.
The function will attempt to create the plot with proper width and height,
but these can be over-written by providing them to the function.
The function will save the plot in the location provided through the
gpro_file_location parameter."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TimeSeriesAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/TS_analysis_PBMC.html">TimeSeries Analysis – PBMC example run</a>
    </li>
    <li>
      <a href="../articles/TS_analysis_celegans.html">TimeSeries Analysis – C.elegans example run</a>
    </li>
    <li>
      <a href="../articles/TS_analysis_murine.html">TimeSeries Analysis – Murine example run</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Ylefol/TimeSeriesAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create gprofiler dotplots</h1>
    <small class="dont-index">Source: <a href="https://github.com/Ylefol/TimeSeriesAnalysis/blob/HEAD/R/MDS_GO_results_functions.R" class="external-link"><code>R/MDS_GO_results_functions.R</code></a></small>
    <div class="hidden name"><code>GO_dotplot_wrapper.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Wrapper function for the creation of a dotplot which summarises the gprofiler findings
for a specified ontology.</p>
<p>The function reads the results for a location. It will automatically search for the
'gprofiler_results/data_files' folder in the procided location. This is should be located
in the TS_results folder, but can change based on the user.</p>
<p>The function will attempt to create the plot with proper width and height,
but these can be over-written by providing them to the function.</p>
<p>The function will save the plot in the location provided through the
gpro_file_location parameter.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GO_dotplot_wrapper</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">file_loc</span>,</span>
<span>  <span class="va">target_ontology</span>,</span>
<span>  <span class="va">top_n</span>,</span>
<span>  custom_width <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  custom_height <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return_plot <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A time series object</p></dd>


<dt>file_loc</dt>
<dd><p>The location where the files are to be saved</p></dd>


<dt>target_ontology</dt>
<dd><p>The targeted ontology ex: 'REAC' or 'GO:BP'</p></dd>


<dt>top_n</dt>
<dd><p>The number of top GOs to plot per cluster</p></dd>


<dt>custom_width</dt>
<dd><p>A custom value for the width of the plot</p></dd>


<dt>custom_height</dt>
<dd><p>A custom value for the height of the plot</p></dd>


<dt>return_plot</dt>
<dd><p>Boolean indicating if the plot should be returned</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>if specified, the function will return the ggplot2 object for the dotplot</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">TS_object</span><span class="op">&lt;-</span><span class="fu"><a href="create_example_object_for_R.html">create_example_object_for_R</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">TS_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize_timeSeries_with_deseq2.html">normalize_timeSeries_with_deseq2</a></span><span class="op">(</span>time_object<span class="op">=</span><span class="va">TS_object</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> converting counts to integer mode</span>
<span class="r-in"><span><span class="co">#Perform conditional differential gene expression analysis</span></span></span>
<span class="r-in"><span><span class="va">TS_object</span><span class="op">&lt;-</span><span class="fu"><a href="conditional_DE_wrapper.html">conditional_DE_wrapper</a></span><span class="op">(</span><span class="va">TS_object</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using pre-existing size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fitting model and testing</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using pre-existing size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fitting model and testing</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using pre-existing size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fitting model and testing</span>
<span class="r-in"><span><span class="co">#Extract genes for PART clustering based on defined log(2)foldChange threshold</span></span></span>
<span class="r-in"><span><span class="va">signi_genes</span><span class="op">&lt;-</span><span class="fu"><a href="select_genes_with_l2fc.html">select_genes_with_l2fc</a></span><span class="op">(</span><span class="va">TS_object</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Use all samples, but implement a custom order. In this case it is reversed</span></span></span>
<span class="r-in"><span><span class="va">sample_data</span><span class="op">&lt;-</span><span class="fu"><a href="exp_sample_data.html">exp_sample_data</a></span><span class="op">(</span><span class="va">TS_object</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">samps_2</span><span class="op">&lt;-</span><span class="va">sample_data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sample_data</span><span class="op">$</span><span class="va">group</span><span class="op">==</span><span class="va">TS_object</span><span class="op">@</span><span class="va">group_names</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">samps_1</span><span class="op">&lt;-</span><span class="va">sample_data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sample_data</span><span class="op">$</span><span class="va">group</span><span class="op">==</span><span class="va">TS_object</span><span class="op">@</span><span class="va">group_names</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Create the matrix that will be used for PART clustering</span></span></span>
<span class="r-in"><span><span class="va">TS_object</span><span class="op">&lt;-</span><span class="fu"><a href="prep_counts_for_PART.html">prep_counts_for_PART</a></span><span class="op">(</span>object<span class="op">=</span><span class="va">TS_object</span>,target_genes<span class="op">=</span><span class="va">signi_genes</span>,scale<span class="op">=</span><span class="cn">TRUE</span>,target_samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">samps_2</span>,<span class="va">samps_1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">TS_object</span><span class="op">&lt;-</span><span class="fu"><a href="compute_PART.html">compute_PART</a></span><span class="op">(</span><span class="va">TS_object</span>,part_recursion<span class="op">=</span><span class="fl">10</span>,part_min_clust<span class="op">=</span><span class="fl">10</span>,dist_param<span class="op">=</span><span class="st">"euclidean"</span>, hclust_param<span class="op">=</span><span class="st">"average"</span>,vignette_run<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">TS_object</span><span class="op">&lt;-</span><span class="fu"><a href="run_gprofiler_PART_clusters.html">run_gprofiler_PART_clusters</a></span><span class="op">(</span><span class="va">TS_object</span>,vignette_run<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co">#Run the gprofiler analysis</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> running Gprofiler on PART clusters</span>
<span class="r-in"><span><span class="co">#Set to NULL to not save results</span></span></span>
<span class="r-in"><span><span class="va">gpro_res</span><span class="op">&lt;-</span><span class="fu"><a href="gprofiler_cluster_analysis.html">gprofiler_cluster_analysis</a></span><span class="op">(</span><span class="va">TS_object</span>,<span class="st">'GO:BP'</span>,save_path<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpro_dotplot</span><span class="op">&lt;-</span><span class="fu">GO_dotplot_wrapper</span><span class="op">(</span><span class="va">TS_object</span>,file_loc<span class="op">=</span><span class="cn">NULL</span>,target_ontology<span class="op">=</span><span class="st">'GO:BP'</span>,top_n<span class="op">=</span><span class="fl">10</span>,return_plot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yohan Lefol.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

