<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ TimeSeriesAnalysis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TimeSeriesAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">01.03.01</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Examples

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/TS_analysis_PBMC.html">PBMC analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Ylefol/TimeSeriesAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Celegans_TS_data.html">Celegans_TS_data</a></code> </p>
        </td>
        <td><p>RNAseq counts for Celegans and krill oil timeseries exeriment time series experiment</p></td>
      </tr><tr><td>
          <p><code><a href="DE_plots_vignettes.html">DE_plots_vignettes()</a></code> </p>
        </td>
        <td><p>Creates differential expression results for Vignettes</p></td>
      </tr><tr><td>
          <p><code><a href="DE_using_DESeq2.html">DE_using_DESeq2()</a></code> </p>
        </td>
        <td><p>DE with DESeq2</p></td>
      </tr><tr><td>
          <p><code><a href="DE_using_limma.html">DE_using_limma()</a></code> </p>
        </td>
        <td><p>DE using limma</p></td>
      </tr><tr><td>
          <p><code><a href="GO_dotplot_wrapper.html">GO_dotplot_wrapper()</a></code> </p>
        </td>
        <td><p>Create gprofiler dotplots</p></td>
      </tr><tr><td>
          <p><code><a href="PART_heat_map.html">PART_heat_map()</a></code> </p>
        </td>
        <td><p>Create and plot PART heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="PBMC_TS_data.html">PBMC_TS_data</a></code> </p>
        </td>
        <td><p>RNAseq counts for AID activation cocktail in PBMC time series experiment</p></td>
      </tr><tr><td>
          <p><code><a href="PBMC_pre_loaded.html">PBMC_pre_loaded</a></code> </p>
        </td>
        <td><p>Pre-computed PART and gprofiler results for the PBMC dataset</p></td>
      </tr><tr><td>
          <p><code><a href="PartRec.html">PartRec()</a></code> </p>
        </td>
        <td><p>PartRec</p></td>
      </tr><tr><td>
          <p><code><a href="SS_GO_clusters.html">SS_GO_clusters()</a></code> </p>
        </td>
        <td><p>Find semantic similarity between clustered GOs</p></td>
      </tr><tr><td>
          <p><code><a href="TS_load_example_data.html">TS_load_example_data()</a></code> </p>
        </td>
        <td><p>Add counts and sample data of examples to a TimeSeriesObject</p></td>
      </tr><tr><td>
          <p><code><a href="TimeSeries_Object-class.html">TimeSeries_Object-class</a></code> <code><a href="TimeSeries_Object-class.html">TimeSeries_Object</a></code> </p>
        </td>
        <td><p>The TimeSeries class</p></td>
      </tr><tr><td>
          <p><code><a href="add_exp_data_tximport.html">add_exp_data_tximport()</a></code> </p>
        </td>
        <td><p>Adds experiment data from tximport estimated reads in the form of a SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="add_experiment_data.html">add_experiment_data()</a></code> </p>
        </td>
        <td><p>Adds experiment data in the form of a SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="add_semantic_similarity_data.html">add_semantic_similarity_data()</a></code> </p>
        </td>
        <td><p>Calculate and add semantic similarity to object</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_EB.html">calculate_EB()</a></code> </p>
        </td>
        <td><p>Empirical bayesan analysis</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_and_format_MDS.html">calculate_and_format_MDS()</a></code> </p>
        </td>
        <td><p>Calculate MDS matrix</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_cluster_traj_data.html">calculate_cluster_traj_data()</a></code> </p>
        </td>
        <td><p>Calcualte trajectory of clusters</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_gene_traj_data.html">calculate_gene_traj_data()</a></code> </p>
        </td>
        <td><p>Calculate trajectory of a single gene</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_mean_cluster_traj.html">calculate_mean_cluster_traj()</a></code> </p>
        </td>
        <td><p>Calculate mean trajectory of clusters</p></td>
      </tr><tr><td>
          <p><code><a href="celegans_pre_loaded.html">celegans_pre_loaded</a></code> </p>
        </td>
        <td><p>Pre-computed PART and gprofiler results for the PBMC dataset</p></td>
      </tr><tr><td>
          <p><code><a href="compute_PART.html">compute_PART()</a></code> </p>
        </td>
        <td><p>compute PART clusters</p></td>
      </tr><tr><td>
          <p><code><a href="conditional_DE_wrapper.html">conditional_DE_wrapper()</a></code> </p>
        </td>
        <td><p>Create conditional DE results</p></td>
      </tr><tr><td>
          <p><code><a href="convert_eb_res_to_DE_results.html">convert_eb_res_to_DE_results()</a></code> </p>
        </td>
        <td><p>Format limma DE results</p></td>
      </tr><tr><td>
          <p><code><a href="create_DEG_df.html">create_DEG_df()</a></code> </p>
        </td>
        <td><p>Create summary table for DEGs</p></td>
      </tr><tr><td>
          <p><code><a href="create_DE_data_results.html">create_DE_data_results()</a></code> </p>
        </td>
        <td><p>Create and save DE results</p></td>
      </tr><tr><td>
          <p><code><a href="create_clustered_module_dataframe.html">create_clustered_module_dataframe()</a></code> </p>
        </td>
        <td><p>Format clustered GO data for plotting</p></td>
      </tr><tr><td>
          <p><code><a href="create_conditional_heatmap_matrix.html">create_conditional_heatmap_matrix()</a></code> </p>
        </td>
        <td><p>Create data for conditional heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="create_example_data_for_R.html">create_example_data_for_R()</a></code> </p>
        </td>
        <td><p>Create a subset of count data for R examples (documentation)</p></td>
      </tr><tr><td>
          <p><code><a href="create_example_object_for_R.html">create_example_object_for_R()</a></code> </p>
        </td>
        <td><p>Create a TimeSeries_Object from example data</p></td>
      </tr><tr><td>
          <p><code><a href="create_raw_count_matrix.html">create_raw_count_matrix()</a></code> </p>
        </td>
        <td><p>Prep TS matrix</p></td>
      </tr><tr><td>
          <p><code><a href="create_tables_genes_of_interest_DE.html">create_tables_genes_of_interest_DE()</a></code> </p>
        </td>
        <td><p>Creates tables and plots for genes of interest</p></td>
      </tr><tr><td>
          <p><code><a href="create_temporal_heatmap_matrix.html">create_temporal_heatmap_matrix()</a></code> </p>
        </td>
        <td><p>Create data for temporal heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="custom_gpro_dotplot.html">custom_gpro_dotplot()</a></code> </p>
        </td>
        <td><p>Create dotplot for gprofiler results</p></td>
      </tr><tr><td>
          <p><code><a href="custom_heatmap_wrapper.html">custom_heatmap_wrapper()</a></code> </p>
        </td>
        <td><p>Create heatmap for differential gene expression results</p></td>
      </tr><tr><td>
          <p><code><a href="doHclust.html">doHclust()</a></code> </p>
        </td>
        <td><p>doHclust</p></td>
      </tr><tr><td>
          <p><code><a href="doKmeans.html">doKmeans()</a></code> </p>
        </td>
        <td><p>doKmeans</p></td>
      </tr><tr><td>
          <p><code><a href="dotplot_ancestors.html">dotplot_ancestors()</a></code> </p>
        </td>
        <td><p>Create dotplot for ancestor query results</p></td>
      </tr><tr><td>
          <p><code><a href="exp_matrix.html">exp_matrix()</a></code> </p>
        </td>
        <td><p>Fetches a data matrix from summarized experiments</p></td>
      </tr><tr><td>
          <p><code><a href="exp_sample_data.html">exp_sample_data()</a></code> </p>
        </td>
        <td><p>Fetches the sample data</p></td>
      </tr><tr><td>
          <p><code><a href="findPartition.html">findPartition()</a></code> </p>
        </td>
        <td><p>findPartition</p></td>
      </tr><tr><td>
          <p><code><a href="findW.html">findW()</a></code> </p>
        </td>
        <td><p>findW</p></td>
      </tr><tr><td>
          <p><code><a href="find_clusters_from_termdist.html">find_clusters_from_termdist()</a></code> </p>
        </td>
        <td><p>Cluster GO terms using Wang measurement</p></td>
      </tr><tr><td>
          <p><code><a href="find_merged_color.html">find_merged_color()</a></code> </p>
        </td>
        <td><p>associate color to group for plotting</p></td>
      </tr><tr><td>
          <p><code><a href="find_most_variable_cluster.html">find_most_variable_cluster()</a></code> </p>
        </td>
        <td><p>Identify cluster with the most variability</p></td>
      </tr><tr><td>
          <p><code><a href="find_relation_to_ancestors.html">find_relation_to_ancestors()</a></code> </p>
        </td>
        <td><p>Ancestor querying for GOs</p></td>
      </tr><tr><td>
          <p><code><a href="gap.html">gap()</a></code> </p>
        </td>
        <td><p>gap</p></td>
      </tr><tr><td>
          <p><code><a href="get.threshold.html">get.threshold()</a></code> </p>
        </td>
        <td><p>get.threshold</p></td>
      </tr><tr><td>
          <p><code><a href="getDist.html">getDist()</a></code> </p>
        </td>
        <td><p>getDist</p></td>
      </tr><tr><td>
          <p><code><a href="getPARTlabels.html">getPARTlabels()</a></code> </p>
        </td>
        <td><p>getPARTlabels</p></td>
      </tr><tr><td>
          <p><code><a href="getReferenceW.html">getReferenceW()</a></code> </p>
        </td>
        <td><p>getReferenceW</p></td>
      </tr><tr><td>
          <p><code><a href="gprofiler_cluster_analysis.html">gprofiler_cluster_analysis()</a></code> </p>
        </td>
        <td><p>Plot gprofiler results</p></td>
      </tr><tr><td>
          <p><code><a href="identify_problematic_combs.html">identify_problematic_combs()</a></code> </p>
        </td>
        <td><p>Identify and remove bad combs</p></td>
      </tr><tr><td>
          <p><code><a href="log_transform_l2fc_vect.html">log_transform_l2fc_vect()</a></code> </p>
        </td>
        <td><p>Log transform L2FC/FC values</p></td>
      </tr><tr><td>
          <p><code><a href="maplot_alt.html">maplot_alt()</a></code> </p>
        </td>
        <td><p>Create an MA plot</p></td>
      </tr><tr><td>
          <p><code><a href="merge_duplicate_modules.html">merge_duplicate_modules()</a></code> </p>
        </td>
        <td><p>merge_MDS modules</p></td>
      </tr><tr><td>
          <p><code><a href="murine_TS_data.html">murine_TS_data</a></code> </p>
        </td>
        <td><p>RNAseq counts for AID activation cocktail in a murine timeseries exeriment time series experiment</p></td>
      </tr><tr><td>
          <p><code><a href="murine_pre_loaded.html">murine_pre_loaded</a></code> </p>
        </td>
        <td><p>Pre-computed PART and gprofiler results for the PBMC dataset</p></td>
      </tr><tr><td>
          <p><code><a href="normalize_timeSeries_with_deseq2.html">normalize_timeSeries_with_deseq2()</a></code> </p>
        </td>
        <td><p>Normalize with DESeq2</p></td>
      </tr><tr><td>
          <p><code><a href="part.html">part()</a></code> </p>
        </td>
        <td><p>part</p></td>
      </tr><tr><td>
          <p><code><a href="part_gprofiler_vignettes.html">part_gprofiler_vignettes()</a></code> </p>
        </td>
        <td><p>Gprofiler analysis for Vignettes</p></td>
      </tr><tr><td>
          <p><code><a href="part_load_results_vignettes.html">part_load_results_vignettes()</a></code> </p>
        </td>
        <td><p>PART analysis for Vignettes</p></td>
      </tr><tr><td>
          <p><code><a href="plot_MDS.html">plot_MDS()</a></code> </p>
        </td>
        <td><p>Create MDS plot</p></td>
      </tr><tr><td>
          <p><code><a href="plot_PCA_TS.html">plot_PCA_TS()</a></code> </p>
        </td>
        <td><p>Plot timeseries PCA</p></td>
      </tr><tr><td>
          <p><code><a href="plot_ancestor_clust_MDS.html">plot_ancestor_clust_MDS()</a></code> </p>
        </td>
        <td><p>Plot nearest ancestor MDS (clustered GOs)</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cluster_traj.html">plot_cluster_traj()</a></code> </p>
        </td>
        <td><p>Plot cluster trajectories</p></td>
      </tr><tr><td>
          <p><code><a href="plot_clustered_mds.html">plot_clustered_mds()</a></code> </p>
        </td>
        <td><p>plot clustered MDS</p></td>
      </tr><tr><td>
          <p><code><a href="plot_custom_DE_heatmap.html">plot_custom_DE_heatmap()</a></code> </p>
        </td>
        <td><p>Plots DE heatmaps</p></td>
      </tr><tr><td>
          <p><code><a href="plot_gene_traj_single_tp.html">plot_gene_traj_single_tp()</a></code> </p>
        </td>
        <td><p>Calculate trajectory of a single gene with no timepoints</p></td>
      </tr><tr><td>
          <p><code><a href="plot_single_gene_traj.html">plot_single_gene_traj()</a></code> </p>
        </td>
        <td><p>Plot gene trajectory</p></td>
      </tr><tr><td>
          <p><code><a href="plot_wrapper_DE_results.html">plot_wrapper_DE_results()</a></code> </p>
        </td>
        <td><p>Creates differential expression results</p></td>
      </tr><tr><td>
          <p><code><a href="prep_RNAseq_matrix.html">prep_RNAseq_matrix()</a></code> </p>
        </td>
        <td><p>prep count matrix form RNAseq count files</p></td>
      </tr><tr><td>
          <p><code><a href="prep_counts_for_PART.html">prep_counts_for_PART()</a></code> </p>
        </td>
        <td><p>Format data for PART</p></td>
      </tr><tr><td>
          <p><code><a href="prep_limma_matrix.html">prep_limma_matrix()</a></code> </p>
        </td>
        <td><p>Prepare data matrix from a Elist object</p></td>
      </tr><tr><td>
          <p><code><a href="prep_sample_data.html">prep_sample_data()</a></code> </p>
        </td>
        <td><p>Sample data set-up</p></td>
      </tr><tr><td>
          <p><code><a href="prep_tp_matrix.html">prep_tp_matrix()</a></code> </p>
        </td>
        <td><p>limma timepoint matrix</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_heat_data.html">prepare_heat_data()</a></code> </p>
        </td>
        <td><p>Process data for DE heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_top_annotation_PART_heat.html">prepare_top_annotation_PART_heat()</a></code> </p>
        </td>
        <td><p>Create top annotation for PART heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="process_microarr_dta_limma.html">process_microarr_dta_limma()</a></code> </p>
        </td>
        <td><p>Read and process microarray data using the limma package</p></td>
      </tr><tr><td>
          <p><code><a href="read_gprofiler_results.html">read_gprofiler_results()</a></code> </p>
        </td>
        <td><p>Read gprofiler data from results folder</p></td>
      </tr><tr><td>
          <p><code><a href="run_gprofiler_PART_clusters.html">run_gprofiler_PART_clusters()</a></code> </p>
        </td>
        <td><p>Gprofiler analysis</p></td>
      </tr><tr><td>
          <p><code><a href="select_genes_with_l2fc.html">select_genes_with_l2fc()</a></code> </p>
        </td>
        <td><p>Select genes based on L2FC/FC</p></td>
      </tr><tr><td>
          <p><code><a href="sim.html">sim()</a></code> </p>
        </td>
        <td><p>sim</p></td>
      </tr><tr><td>
          <p><code><a href="temporal_DE_wrapper.html">temporal_DE_wrapper()</a></code> </p>
        </td>
        <td><p>Create temporal DE results</p></td>
      </tr><tr><td>
          <p><code><a href="volcanoplot_alt.html">volcanoplot_alt()</a></code> </p>
        </td>
        <td><p>Create a volcano plot</p></td>
      </tr><tr><td>
          <p><code><a href="wrapper_MDS_and_MDS_clusters.html">wrapper_MDS_and_MDS_clusters()</a></code> </p>
        </td>
        <td><p>Create MDS and nearest ancestor MDS plots</p></td>
      </tr><tr><td>
          <p><code><a href="wrapper_ancestor_curation_plots.html">wrapper_ancestor_curation_plots()</a></code> </p>
        </td>
        <td><p>Create ancestor query plots</p></td>
      </tr><tr><td>
          <p><code><a href="wrapper_cluster_trajectory.html">wrapper_cluster_trajectory()</a></code> </p>
        </td>
        <td><p>Plot cluster trajectory for all PART clusters</p></td>
      </tr><tr><td>
          <p><code><a href="write_example_data_to_dir.html">write_example_data_to_dir()</a></code> </p>
        </td>
        <td><p>Prep example data in directory</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yohan Lefol.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

