<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Creates differential expression results — plot_wrapper_DE_results • TimeSeriesAnalysis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Creates differential expression results — plot_wrapper_DE_results"><meta property="og:description" content="The function will create a folder called DE_results_(conditonal or temporal),
it will then create a summary heatmap of all the differential expression experiments
done in that type.
The function will then iterate over the differential gene expression experiments done in
either DE_type (conditional or temporal) and creates two csv files, one containing
the raw differential expression data and the other is a subset of the raw data
containing only the significant genes
The function also creates a volcano plot and MA plot for each experiment
The individual differential expression results are stored in folders which
carry the name of the experiment"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TimeSeriesAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">01.03.01</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Examples

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/TS_analysis_PBMC.html">PBMC analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Ylefol/TimeSeriesAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Creates differential expression results</h1>
    <small class="dont-index">Source: <a href="https://github.com/Ylefol/TimeSeriesAnalysis/blob/HEAD/R/DE_PART_results_functions.R" class="external-link"><code>R/DE_PART_results_functions.R</code></a></small>
    <div class="hidden name"><code>plot_wrapper_DE_results.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function will create a folder called DE_results_(conditonal or temporal),
it will then create a summary heatmap of all the differential expression experiments
done in that type.</p>
<p>The function will then iterate over the differential gene expression experiments done in
either DE_type (conditional or temporal) and creates two csv files, one containing
the raw differential expression data and the other is a subset of the raw data
containing only the significant genes
The function also creates a volcano plot and MA plot for each experiment</p>
<p>The individual differential expression results are stored in folders which
carry the name of the experiment</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_wrapper_DE_results</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">DE_type</span>,</span>
<span>  genes_of_interest <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  results_folder <span class="op">=</span> <span class="st">"TS_results/"</span>,</span>
<span>  adjust_missing_temp_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  do_SVGs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  do_heatmap <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>A timeseries object</p></dd>


<dt id="arg-de-type">DE_type<a class="anchor" aria-label="anchor" href="#arg-de-type"></a></dt>
<dd><p>The type of differential gene expression done that the function
will use to create results. Either conditional or temporal</p></dd>


<dt id="arg-genes-of-interest">genes_of_interest<a class="anchor" aria-label="anchor" href="#arg-genes-of-interest"></a></dt>
<dd><p>A vector of genes of interest, can be empty (c())</p></dd>


<dt id="arg-results-folder">results_folder<a class="anchor" aria-label="anchor" href="#arg-results-folder"></a></dt>
<dd><p>The main folder where the results will be stored</p></dd>


<dt id="arg-adjust-missing-temp-samples">adjust_missing_temp_samples<a class="anchor" aria-label="anchor" href="#arg-adjust-missing-temp-samples"></a></dt>
<dd><p>Boolean indicating if missing temporal samples should be
compensated for (NA introduced). Currently, there is no reason to have this set to
False.</p></dd>


<dt id="arg-do-svgs">do_SVGs<a class="anchor" aria-label="anchor" href="#arg-do-svgs"></a></dt>
<dd><p>Boolean indicating if SVGs should be saved for the heatmaps</p></dd>


<dt id="arg-do-heatmap">do_heatmap<a class="anchor" aria-label="anchor" href="#arg-do-heatmap"></a></dt>
<dd><p>Boolean indicating if the custom heatmap should be performed</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>None</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yohan Lefol.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

