<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plots DE heatmaps — plot_custom_DE_heatmap • TimeSeriesAnalysis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plots DE heatmaps — plot_custom_DE_heatmap"><meta property="og:description" content="The function is intended to be called within the custom_heatmap_wrapper
The function creates a segmented heatmap where rows are samples/patients and columns
are genes. The heatmap is column segmented based on the number of differential expression
analyses contained within it. A log2foldChange histogram is attached to the bottom
of the heatmap to indicate significance of the genes being visualized.
The column/segment legend is given at the bottom of the heatmap with the number
of genes in each group being also provided.
The gradient legend of the values within the heatmap is given on the right hand side.
The heatmap is saved in both png and svg format."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TimeSeriesAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">01.03.01</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Examples

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/TS_analysis_PBMC.html">PBMC analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Ylefol/TimeSeriesAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plots DE heatmaps</h1>
    <small class="dont-index">Source: <a href="https://github.com/Ylefol/TimeSeriesAnalysis/blob/HEAD/R/DE_PART_results_functions.R" class="external-link"><code>R/DE_PART_results_functions.R</code></a></small>
    <div class="hidden name"><code>plot_custom_DE_heatmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function is intended to be called within the <code>custom_heatmap_wrapper</code></p>
<p>The function creates a segmented heatmap where rows are samples/patients and columns
are genes. The heatmap is column segmented based on the number of differential expression
analyses contained within it. A log2foldChange histogram is attached to the bottom
of the heatmap to indicate significance of the genes being visualized.</p>
<p>The column/segment legend is given at the bottom of the heatmap with the number
of genes in each group being also provided.</p>
<p>The gradient legend of the values within the heatmap is given on the right hand side.</p>
<p>The heatmap is saved in both png and svg format.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_custom_DE_heatmap</span><span class="op">(</span></span>
<span>  <span class="va">heat_mat</span>,</span>
<span>  <span class="va">col_split</span>,</span>
<span>  <span class="va">row_splits</span>,</span>
<span>  <span class="va">l2fc_col</span>,</span>
<span>  <span class="va">log_transform</span>,</span>
<span>  <span class="va">color_vect</span>,</span>
<span>  legend_value <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  plot_file_name <span class="op">=</span> <span class="st">"custom_heatmap"</span>,</span>
<span>  custom_width <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  custom_height <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  do_SVG <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-heat-mat">heat_mat<a class="anchor" aria-label="anchor" href="#arg-heat-mat"></a></dt>
<dd><p>The main count matrix of the heatmap</p></dd>


<dt id="arg-col-split">col_split<a class="anchor" aria-label="anchor" href="#arg-col-split"></a></dt>
<dd><p>The named vector used to split the columns into it's segments</p></dd>


<dt id="arg-row-splits">row_splits<a class="anchor" aria-label="anchor" href="#arg-row-splits"></a></dt>
<dd><p>The named vector used to split the rows into the two groups</p></dd>


<dt id="arg-l-fc-col">l2fc_col<a class="anchor" aria-label="anchor" href="#arg-l-fc-col"></a></dt>
<dd><p>The log2foldchange vector used to create the histogram</p></dd>


<dt id="arg-log-transform">log_transform<a class="anchor" aria-label="anchor" href="#arg-log-transform"></a></dt>
<dd><p>boolean indicating if the results were log transformed or not</p></dd>


<dt id="arg-color-vect">color_vect<a class="anchor" aria-label="anchor" href="#arg-color-vect"></a></dt>
<dd><p>A vector for the colors to be used in the heatmap</p></dd>


<dt id="arg-legend-value">legend_value<a class="anchor" aria-label="anchor" href="#arg-legend-value"></a></dt>
<dd><p>string indicating the type of value used
(counts for RNAseq, intensity value for microarray)</p></dd>


<dt id="arg-plot-file-name">plot_file_name<a class="anchor" aria-label="anchor" href="#arg-plot-file-name"></a></dt>
<dd><p>The name given to the saved heatmap</p></dd>


<dt id="arg-custom-width">custom_width<a class="anchor" aria-label="anchor" href="#arg-custom-width"></a></dt>
<dd><p>The width of the heatmap</p></dd>


<dt id="arg-custom-height">custom_height<a class="anchor" aria-label="anchor" href="#arg-custom-height"></a></dt>
<dd><p>The height of the heatmap</p></dd>


<dt id="arg-do-svg">do_SVG<a class="anchor" aria-label="anchor" href="#arg-do-svg"></a></dt>
<dd><p>Boolean to check if the SVG file should be created or not</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>none or heatmap plot</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">TS_object</span><span class="op">&lt;-</span><span class="fu"><a href="create_example_object_for_R.html">create_example_object_for_R</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">TS_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize_timeSeries_with_deseq2.html">normalize_timeSeries_with_deseq2</a></span><span class="op">(</span>time_object<span class="op">=</span><span class="va">TS_object</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> converting counts to integer mode</span>
<span class="r-in"><span><span class="co">#Perform conditional differential gene expression analysis</span></span></span>
<span class="r-in"><span><span class="va">TS_object</span><span class="op">&lt;-</span><span class="fu"><a href="conditional_DE_wrapper.html">conditional_DE_wrapper</a></span><span class="op">(</span><span class="va">TS_object</span>,vignette_run<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">heat_dta</span><span class="op">&lt;-</span><span class="fu"><a href="create_conditional_heatmap_matrix.html">create_conditional_heatmap_matrix</a></span><span class="op">(</span><span class="va">TS_object</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">heat_dta</span><span class="op">&lt;-</span><span class="fu"><a href="prepare_heat_data.html">prepare_heat_data</a></span><span class="op">(</span><span class="va">heat_dta</span>,log_transform<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">log_l2fc</span><span class="op">&lt;-</span><span class="fu"><a href="log_transform_l2fc_vect.html">log_transform_l2fc_vect</a></span><span class="op">(</span><span class="va">heat_dta</span><span class="op">[[</span><span class="st">'l2fc_vector'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">heat_plot</span><span class="op">&lt;-</span> <span class="fu">plot_custom_DE_heatmap</span><span class="op">(</span><span class="va">heat_dta</span><span class="op">[[</span><span class="st">'heat_matrix'</span><span class="op">]</span><span class="op">]</span>,<span class="va">heat_dta</span><span class="op">[[</span><span class="st">'region_split'</span><span class="op">]</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                                   <span class="va">heat_dta</span><span class="op">[[</span><span class="st">'group_split'</span><span class="op">]</span><span class="op">]</span>,<span class="va">log_l2fc</span>,log_transform <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                                   legend_value<span class="op">=</span><span class="st">'counts'</span>, plot_file_name <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                   do_SVG<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in plot_custom_DE_heatmap(heat_dta[["heat_matrix"]], heat_dta[["region_split"]],     heat_dta[["group_split"]], log_l2fc, log_transform = TRUE,     legend_value = "counts", plot_file_name = NULL, do_SVG = FALSE):</span> argument "color_vect" is missing, with no default</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yohan Lefol.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

